keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -3,703 +3,903 @@
+
+When a user deletes a BucketAccess, the corresponding secret/serviceaccount are also deleted. If a pod has that secret mounted when delete is called, then a finalizer on the secret will prevent it from being deleted. Instead, the deletionTimestamp will be set on the secret. In this way, access to a Bucket is preserved until the application pod dies. 
+
+When an admin deletes any of the class objects, it does not affect existing Buckets as fields from the class objects are copied into the Buckets during creation. 
+
+If a Bucket is manually deleted by an admin, then a finalizer on the Bucket prevents it from being deleted until the binding BucketClaim is deleted. 
+
+# Usability
+
+## Self Service
+
+Self service is easily possible with the current design as both the BucketRequest and BucketClaim resources are namespace scoped, and users need not have admin privileges to create, modify and delete them.
+
+The only admin steps are creation of class objects(BucketClass, BucketAccessClass) and Bucket imports. The creatio of class object is no different from requiring a StorageClass for provisioning PVCs. It is a well-understood pattern among kubernetes users. Importing a Bucket requires special permissions because its lifecycle is not managed by COSI, and special care needs to be taken to prevent clones, accidental deletions and other mishaps (for instance, setting the deletion policy to Delete).
+
+## Mutating Buckets
+
+As of the current design of COSI, mutating bucket properties is not supported. However, the current design does not prevent us from supporting it in the future. Mutable properties include encryption, object lifecycle configuration, replication configuration etc. These properties will be supported in future versions along with the capability to mutate them.
+
+These properties will be specified in the BucketRequest and follow the same pattern of events as Bucket creation. i.e. Bucket API object will be updated to reflect the properties set in BucketRequest, and then a controller will pick-up these changes and call the appropriate APIs to reflect them in the backend.
+
+# Object Lifecycle
+
+The following resources are managed by admins
+
+- Bucket in case of brownfield buckets 
+- BucketClass
+- BucketAccessClass
+
+The following resources are managed by a user. These are created with a reference to their corresponding class objects:
+
+- BucketClaim -> BucketClass
+- BucketAccess -> Bucket, BucketClaim, BucketAccessClass
+
+The COSI controller responds by creating the following objects
+
+- BucketRequest <-> Bucket (1:1)
+- BucketClaim -> Bucket
+
+Notes:
 
-message ProvisionerCreateBucketResponse {
-    // Intentionally left blank
+ - There are **NO** cycles in the relationship graph of the above mentioned API objects.
+ - Mutations are not supported in the API.
+
+# COSI API Reference
+
+## Bucket
+
+Resource to represent a Bucket in OSP. Buckets are cluster-scoped.
+
+```yaml
```

Comment:
These blocks are roughly Go and render more consistently that way on GitHub. Same suggestion for all the API blocks.

```suggestion
```go
```